### DISCUSSION TITLE
Setting Callback Event for Notification

### DISCUSSION THREAD

**Discussion by bagusnl** [HavenDV/H.NotifyIcon#163] [Q&A] [answered]
Hi!

I want to know how one can call a function when notification is clicked by the user.

Thanks

**Comment by HavenDV** [comment]
![image](https://github.com/user-attachments/assets/fb0fa055-f7e6-48a3-bde0-a900eb2a8dbc)
I don't remember exactly, but I think it was these events.
Not the best inherited names

**Reply by bagusnl** [reply]
![image](https://github.com/user-attachments/assets/ba577e63-841b-430d-989a-f53410279480)

There doesn't seem to be those routed events found on my instance

**Reply by HavenDV** [reply, marked as answer]
What platform are you on? This is only available for WPF. There is no such thing as RoutedEvents in WinUI. I thought about adding regular events to replace RoutedEvents for the WinUI platform for DependencyPropertyGenerator, but never got around to it.
You can still access this from WinUI using the TaskbarIcon.TrayIcon.MessageWindow events, I made this available in the latest version.
![image](https://github.com/user-attachments/assets/25f37202-9e61-48d3-9aed-b8aafa688b28)
You need MouseEventReceived and BalloonToolTipChanged events

**Reply by bagusnl** [reply]
Sorry, I forgot to state my platform. But yeah its WinUI. With the new version, now I can invoke the event. This is what I did:

In constructor:
```csharp
CollapseTaskbar.TrayIcon.MessageWindow.BalloonToolTipChanged += BalloonChangedEvent;
```

```csharp
private void BalloonChangedEvent(object o, MessageWindow.BalloonToolTipChangedEventArgs args)
{
    CollapseTaskbar.TrayIcon.MessageWindow.MouseEventReceived += NotificationOnMouseEventReceived;
}

private bool _mouseEventProcessing;
private async void NotificationOnMouseEventReceived(object o, MessageWindow.MouseEventReceivedEventArgs args)
{
    if (_mouseEventProcessing) return;
    if (args.MouseEvent != MouseEvent.BalloonToolTipClicked) return;
    
    _mouseEventProcessing = true;
    BringToForeground();
    await Task.Delay(250);
    _mouseEventProcessing = false;
}
```

The only minor issue I have now is that the event is not invoked after the notification went into the Notification tray. But I think I can live without it.

Thank you for the help!

**Reply by HavenDV** [reply]
Also take a look to https://learn.microsoft.com/en-us/windows/apps/design/shell/tiles-and-notifications/adaptive-interactive-toasts?tabs=appsdk
I think it's more modern way to do notifications in WinUI

**Reply by bagusnl** [reply]
Thanks for the heads up on that one, but unfortunately since our app is running as admin, the notification using that method doesn't work
Ref https://github.com/microsoft/WindowsAppSDK/issues/3815#issuecomment-2076812250
